### DISCUSSION TITLE
Async PubSub listeners?

### DISCUSSION THREAD

**Discussion by mikeyk** [alisaifee/coredis#124] [Ideas]
Hi @alisaifee! 

Currently we use one `PubSub` object per user connection, and then use `listen()` to handle incoming responses on just that one subscribed channel. We'd like to move to having one PubSub per `uvicorn` worker instead. To do that, however, we need to have the registered handlers be async, since they need to do work like write back to the client websockets.

A proposal would be:
1. Make [handle_message](https://github.com/alisaifee/coredis/blob/master/coredis/commands/pubsub.py#LL324C18-L324C18) an async function
2. When grabbing handlers from the `self.patterns` or `self.handlers` map, check if they are coroutines or sync functions; if sync, just call regularly, otherwise `await` them. (can be something like Starlette's [is_async_callable](https://github.com/encode/starlette/blob/master/starlette/_utils.py#L6) )

What do you think? Only downside would be a slightly more expensive call to `handle_message` since making it async has a small amount of overhead. Happy to take this on if it sounds good to you.

**Comment by alisaifee** [comment]
On first glance this seems perfectly reasonable - and yes please go ahead and put up a PR!

**Reply by mikeyk** [reply]
Great, thanks; here's the PR: https://github.com/alisaifee/coredis/pull/125

**Comment by alisaifee** [comment]
Available now in [4.10.0](https://github.com/alisaifee/coredis/releases/tag/4.10.0)
